### ISSUE TITLE
Reading pixels of an image from two threads causes errors

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#35539]
Godot 3.2 rc3

I need to read pixels of an image from multiple threads at once. My expectation was that `lock()` would trigger copy-on-write if it's already locked by another thread (since it's actually Write lock inside). But instead, it produces errors:
```
ERROR: Image must be locked with 'lock()' before using get_pixel().
   At: core/image.cpp:2389
```
I originally had this issue in C++ while developping a module, so I had a peek at the state of the Write struct: the pointer to pixels ended up null.

Reproduction project:
[ImageReadInTwoThreads.zip](https://github.com/godotengine/godot/files/4111164/ImageReadInTwoThreads.zip)

**Comment by nathanwfranke** [comment]
As a workaround, consider locking the image in ready and unlocking it in _exit_tree or something

As for the issue, chances are the threads make your execution desynced like this

```
Thread 1, get pixel 98
Thread 2, get pixel 40
Thread 2, get pixel 41    # Due to multithreading you may get inconsistent execution
Thread 1, get pixel 99
Thread 2, get pixel 42
Thread 1, unlock image
Thread 2, get pixel 43    # ERROR
```

**Comment by akien-mga** [comment]
As @RandomShaper commented on IRC:
```
03:55 <RandomShaper>  lock()  is used to lock the image data for writing.
03:56 <RandomShaper> Precisely that's for protecting against concurrent access from multiple threads.
```

The only issue here is that the error is misleading, it should be telling you that the image is already locked by another thread.

**Comment by Zylann** [comment]
So the whole `PoolVector::Write` copy-on-write system (which DOES call `_copy_on_write`) is only to write errors in the console if I want to write from two threads and NOT to transparently copy-on-write in such situations? Indeed I'm confused... 

Locking beforehand isn't ideal either because the image can't be changed anymore while it is read by other threads. What I did instead is to duplicate the image each time I give it to a thread.

Also, due to my use case, this would be another point for https://github.com/godotengine/godot/issues/13934#issuecomment-569655494

**Comment by Zylann** [comment]
When investigating more, I think I know what is really happening.

If I were to send a `PoolVector<uint8_t>` (even by reference) to two threads and use `Write` on them both at the same time, no problem occurs.

However, here I'm passing an `Image`, which in turns contains the PoolVector. That means while the PoolVector itself might be thread-safe regarding calls to `write()`, the `Write` struct stored in `Image` is not protected against threads. So there is a chance that it ends up `null` because two threads are trying to write the result of `write()` into it.
In other words it's just a resource member being accessed by two threads and has nothing to do with CoW...

So duplicating the `Image` resource is actually the right thing to do: it will copy the Image object but the `data` will remain shared until the first `write` occurs.
This may be useful to know when dealing with Images in threads, as the `lock` function doesn't protect at all against this.

**Comment by albinaask** [comment]
I can't but concur, I tested the same method in GDScript and it produces the same result

**Comment by MaaaxiKing** [comment]
I also cannot read correctly or write to pixels.
